강원대_이진호_Step2 기능 구현

treasure-sky opened the pull request:
## step2 구현 기능

- [x] 공통 페이지네이션 응답 dto 구현
- [x] Item 페이지네이션 구현
- [x] Wishlist 페이지네이션 구현
- [x] 페이지네이션 테스트 작성

## 고려사항
### 페이지네이션 공통 응답 dto

- 페이지네이션을 구현한다면 기존의 dto에는 페이지 정보를 담을 수 없어서 새로운 dto가 필요하였습니다.
- 각 도메인 별로 페이지네이션용 dto를 만들기에는 페이지 정보가 대부분 중복되는 내용이었습니다.
- 따라서, 페이지네이션용 공통 응답 dto를 구현하였습니다.
- Page 객체를 그대로 반환하는 것도 가능하긴 하지만, 필요한 정보만 노출 시키고 스프링 프레임워크에 API가 종속되지 않게 하기 위해 dto로 분리하는 것이 옳다고 생각하였습니다.

## 질문
**문제상황**

- 사용자로부터 조회할 페이지의 인덱스를 받아올 때, 가장 첫 페이지의 인덱스를 1로 하려고 하였습니다.
- 일반적으로 페이지 수가 0이 아닌 1부터 시작하기 때문입니다.
- 만약 첫 페이지의 인덱스를 0으로 한다면, 프론트 측에서 화면에 보여줄 때 +1을 한다든가 요청시 -1을 해야하므로 휴먼에러가 발생할 수 있습니다.
- 하지만 spring data JPA 에서는 첫 페이지의 인덱스가 0으로 처리됩니다.
- 이를 해결하기 위해 service 레이어에서 pageable을 생성할 때 사용자의 요청 page 인덱스에서 -1 을 하도록 하였습니다.
- 하지만, 응답으로 나가는 totalPages 는 사용자의 요청값이 아닌 0-based 인덱스로 설정이 되었습니다.
- 공통응답dto에 존재하는 아래와 같은 메서드를 사용하여 Page 객체에서 바로 데이터를 가져왔기 때문입니다.
    
    ```java
    public static <T> PageResponseDto<T> from(Page<T> page) {
            return new PageResponseDto<>(
                page.getContent(),
                page.getNumber(), // <- Spring Data JPA 메서드가 반환하는 page 객체에서 변형하지 않고 바로 가져옴
                page.getSize(),
                page.getTotalElements(),
                page.getTotalPages()
            );
        }
    ```
    
- 이를 해결하기 위해 위 메서드에서 `page.getNumber() - 1` 로 수정하는 방법을 생각하였습니다.
- 하지만 이는 service의 비즈니스 로직이 공통 dto에 영향을 주게 되어 의존성이 생기게 됩니다.
- 다른 API에서 1-based가 아닌 0-based 인덱스를 반환해야한다면 문제가 발생할 수 있습니다.

**해결방안**

1. static 메서드를 제거하고 service 단에서 dto 변환을 전부 책임지게 하기
2. 0-based, 1-based static 메서드를 분리해서 사용하기
    
    ```java
    public static <T> PageResponseDto<T> fromOneBasedPage(
    	Page<T> page,
    	int originalPageNumber
    ) {
            return new PageResponseDto<>(
                page.getContent(),
                originalPageNumber,    // 원래 요청 페이지 번호
                page.getSize(),
                page.getTotalElements(),
                page.getTotalPages()
            );
        }
    ```
    

**최종구현**

1. static 메서드를 제거하고 service 단에서 dto 변환을 전부 책임지게 하기

**이유**

1. 공통 dto의 책임을 최소화하여 의존성을 줄일 수 있음.
2. 0-based, 1-based가 아닌 다른 방식을 사용하는 경우 확장하기 쉬움

**위 상황에서 제가 고려한 사항들에 대해 어떻게 생각하시나요?**
**다른 좋은 해결방법이 있을까요?**

treasure-sky committed:
refactor: 불필요한 메서드 삭제

현재 사용되지 않는 setter나 생성자를 제거하였음.

treasure-sky committed:
docs: step2 구현 기능 작성

treasure-sky committed:
feat: 공통 페이지네이션 응답 dto 구현

treasure-sky committed:
feat: Item 페이지네이션 구현

- 첫페이지 인덱스에 따라 응답이 달라지므로 service에서 이를 처리하게 하였음.

- 따라서, 필요 없는 static 메서드를 제거하였음.

treasure-sky committed:
feat: 정렬 파라미터에 대한 예외처리

- 정렬 기준과 정렬 방향에 대한 파라미터가 올바른지 확인하고 예외를 던지게 하였음.

treasure-sky committed:
fix: 페이지네이션 파라미터로 createdAt 파라미터 제한

- ItemEntity에는 createdAt 파라미터가 없으므로 이를 허용된 파라미터 set에서 제거하였음.

treasure-sky committed:
feat: admin 페이지에 페이지네이션 기능 반영

treasure-sky committed:
feat: wishlist 페이지네이션 구현

treasure-sky committed:
fix: 페이지네이션 시에 잘못된 정렬 기준을 사용하는 오류 수정

- 기존에는 DB 칼럼 필드를 기준으로 JPA repository 메서드를 호출하였음.

- 하지만, DB 칼럼 필드가 아닌 객체 필드의 이름을 사용하여야 함.

- 따라서 이를 변경하였음.

treasure-sky committed:
feat: 예외 메세지 가독성 향상

treasure-sky committed:
feat: 예외 메세지 내용 추가

treasure-sky committed:
feat: 파라미터 validation시 발생하는 예외 처리 구현

- 쿼리 파라미터 validation 시에 발생하는 ConstraintViolationException을 처리하는 전역 예외 처리 메서드를 추가하였음.

treasure-sky committed:
fix: 쿼리 파라미터에 대한 검증이 이루어지지 않는 버그 수정

treasure-sky committed:
test: 위시리스트 페이지네이션 테스트 작성

hjjae2 commented on src/main/java/gift/admin/AdminItemController.java:
```diff
@@ -29,9 +30,19 @@ public AdminItemController(ItemService itemService) {
-        model.addAttribute("items", items);
+    public String adminItemPage(
+        Model model,
+        @RequestParam(defaultValue = "1") int page,
+        @RequestParam(defaultValue = "10") int size,
+        @RequestParam(defaultValue = "id") String sortBy,
+        @RequestParam(defaultValue = "asc") String direction
```
> _스프링 프레임워크에 API가 종속되지 않게 하기 위해 dto로 분리하는 것이 옳다고 생각하였습니다._

특정 프레임워크나 기술, 라이브러리에 의존성을 갖지 않기 위해 직접 구현해주신 부분이 인상깊네요. 좋은 방향이라고 생각합니다. 👍

hjjae2 commented on src/main/java/gift/item/service/ItemService.java:
```diff
@@ -32,22 +35,34 @@ public ItemResponseDto findItem(Long itemId) {
-            itemResponseDtos.add(dto);
-        }
+    public PageResponseDto<ItemResponseDto> findAll(
+        int page,
+        int size,
+        String sortBy,
+        String direction
+    ) {
+        Sort sort = direction.equalsIgnoreCase("asc") ?
+            Sort.by(sortBy).ascending() :
+            Sort.by(sortBy).descending();
+
+        Pageable pageable = PageRequest.of(page - 1, size, sort);
```
인자로 들어온 `page`, `size`, `sortBy`, `direction` 이 의도한 값(예: 양수, `asc|desc` 중 하나 등) 임을 보장할 수 있나요?

treasure-sky replied:
네. controller 단에서 입력값에 대한 검증을 하고 있어서 service 단에 주어지는 값들은 의도한 값임을 보장할 수 있습니다.

```java
private static final Set<String> ALLOWED_SORT_FIELDS =
        Set.of("member", "item", "id", "createdAt");

public ResponseEntity<PageResponseDto<WishlistResponseDto>> findAll(
    @RequestParam(defaultValue = "1")
    @Positive(message = "페이지 인덱스는 양수이어야 합니다.")
    int page,
    @RequestParam(defaultValue = "10")
    @Positive(message = "페이지 사이즈는 양수이어야 합니다.")
    int size,
    @RequestParam(defaultValue = "createdAt") String sortBy,
    @RequestParam(defaultValue = "desc") String direction,
    @LoginMember AuthenticatedMember member
) {
    if (!direction.equalsIgnoreCase("asc") &&
        !direction.equalsIgnoreCase("desc")) {
        throw new InvalidSortDirectionException(direction);
    }
    if (!ALLOWED_SORT_FIELDS.contains(sortBy)) {
        throw new InvalidSortByException(sortBy);
    }
```

hjjae2 replied:
인자는 `int`, `String` 타입인데 어떻게 보장할 수 있을까요? 

말씀주신 건 앞단에서의 검증을 전제하고 있는데, 이 전제가 없을 수도 있지 않나요? 예를 들면, 다른 개발자가 다른 코드에서 이 메서드를 사용한다고 가정하면 똑같이 보장할 수 있을지에 대한 물음이에요.

treasure-sky replied:
service 단에서도 추가적으로 검증을 한다면 이를 해결할 수 있을 거 같습니다.
이를 좀 더 편하게 하기 위해서 dto를 사용할 수 있을 거 같은데요.
검증 어노테이션이 붙은 dto를 사용해서 요청을 받고 이를 서비스 단까지 넘겨준다면 컨트롤러든 서비스든 @Valid 만 붙여서 검증을 할 수 있을 거 같습니다.

hjjae2 replied:
좋네요. 👍 

혹은 각 값을 표현/보장해줄 수 있는 클래스를 만들어 사용해보면 어떨까요? 
예를 들면, `PositiveNumber`, `SortBy`, `Direction` 와 같은 클래스가 생길 수 있겠어요. 
* SortBy, Direction 의 경우 Enum 으로 관리해볼 수도 있겠고 PositiveNumber 의 경우 원시 값을 포장한 클래스([참고](https://tecoble.techcourse.co.kr/post/2020-05-29-wrap-primitive-type/)) 의 형태를 기대하고 의견드립니다.

예시로, 사용처에서의 호출 코드는 다음의 형태가 될 수 있겠어요.

```java
PageResponseDto<ItemResponseDto> pagedDtos = itemService.findAll(
    new Page(page),
    new Size(size),
    SortBy.from(sortBy),
    Direction.from(direction)
);
```

hjjae2 commented on src/main/java/gift/item/service/ItemService.java:
```diff
@@ -32,22 +35,34 @@ public ItemResponseDto findItem(Long itemId) {
-            itemResponseDtos.add(dto);
-        }
+    public PageResponseDto<ItemResponseDto> findAll(
+        int page,
+        int size,
+        String sortBy,
+        String direction
+    ) {
+        Sort sort = direction.equalsIgnoreCase("asc") ?
+            Sort.by(sortBy).ascending() :
+            Sort.by(sortBy).descending();
+
+        Pageable pageable = PageRequest.of(page - 1, size, sort);
```
> _스프링 프레임워크에 API가 종속되지 않게 하기 위해 dto로 분리하는 것이 옳다고 생각하였습니다._

1.

`Pageable` 에 의존하지 않는 것을 의도하신 것이라 이해한 상태에서, 궁금한 점인데요! 

의도하신건 controller 레이어에서만 스프링 프레임워크(`Pageable`) 에 종속되지 않은 것을 의도하신걸까요?
다른 레이어(service, repository)에서는 `Pageable`에 의존하고 있는 것으로 보여서요. 기능 요구사항에 따라 `Pageable` 을 사용할 수 밖에 없었기에, 의존을 가져가게 된걸까요-?

2.

고민해 주신 부분(0-based, 1-based 처리)과 관련해, 바로 위의 질문이 고민을 유발하게 한 근본 지점으로 보이기도 합니다.
애플리케이션 내에서 절반(service, repository)은 0-based(`Pageable` ⭕) 기반으로 작성되어 있고, 절반(controller)은 1-based(`Pageable` ❌) 로 처리하려고 하고 있습니다. 모든 레이어가 하나의 방향(0-based 혹은 1-based)으로 맞춰진다면 이 고민이 해결될 수 있지 않을까요?

treasure-sky replied:
1. 제가 의도한 것은 Page 객체를 직접적으로 반환하게 되면 API 명세가 Page 객체(스프링 프레임워크)에 의존하게 되는 것을 피하는 것이었습니다. 따라서 DTO를 사용하였습니다. 나중에 스프링 프레임워크가 아닌 다른 기술을 사용하더라도 기존에 API를 사용하고 있던 모듈이 정상 작동하도록 하도록 말이죠.
그래서 레이어 별로 Page(Pageable) 객체를 사용할지 안할지는 고려하지 않았습니다.
그리고 쿼리를 직접 구현한다면 굳이 Pageable 객체를 사용하지 않아도 된다고 생각합니다. 하지만, 이번에는 JPA의 기능을 사용해보고자 해당 객체를 사용하였습니다.

2. 네 기준이 모든 레이어에서 통일된다면 해결될 것 같습니다!
Spring Data JPA는 0-based 로 되어있기 때문에 0-based로 통일하는 것은 어렵지 않을 것 같습니다.
하지만, 클라이언트가 이를 알고 프론트 단에서 추가적인 작업이 더 필요할 수 있을 거 같다는 생각을 해서 지금 코드에서는 1-based 로 변환을 해준 것입니다.
그리고 Spring Data JPA에는 0-based, 1-based를 바꾸는 설정은 지원하지 않아서 JPA를 사용한다면 모두 1-based로 바꾸는 것은 어려울 거 같은데 혹시 멘토님의 해결방법이 있을까요??

hjjae2 replied:
> 그리고 Spring Data JPA에는 0-based, 1-based를 바꾸는 설정은 지원하지 않아서 JPA를 사용한다면 모두 1-based로 바꾸는 것은 어려울 거 같은데 혹시 멘토님의 해결방법이 있을까요??

말씀주신 것 처럼 Spring Data JPa, `Pageable` 을 사용한다는 가정 하에, 저희가 직접 코드로 1-based 처리를 해주는 수 밖에 없겠습니다. 

그렇다면 이 처리를 어떻게 관리하기 쉽게 작성할지가 포인트가 되겠는데요. 이에 대한 내용은 [뒤의 코멘트](https://github.com/next-step/spring-gift-enhancement/pull/211#discussion_r2215215322)에서 얘기하게 되겠네요. 🙂

hjjae2 commented on src/main/java/gift/item/service/ItemService.java:
```diff
@@ -32,22 +35,34 @@ public ItemResponseDto findItem(Long itemId) {
+
+        return new PageResponseDto<>(
+            pagedDtos.getContent(),
+            pagedDtos.getNumber() + 1,
```
> static 메서드를 제거하고 service 단에서 dto 변환을 전부 책임지게 하기
> 공통 dto의 책임을 최소화하여 의존성을 줄일 수 있음.
> 0-based, 1-based가 아닌 다른 방식을 사용하는 경우 확장하기 쉬움

고민주신 부분에 대해 이야기 해보아요.

1. service 단에서 처리한다면 모든 service 가 이 로직(pageNumber 에 +1 을 하거나 -1 을 하는 로직)을 알고, 구현해줘야 하지 않을까요? 
2. 만약 DTO 가 0-based, 1-based 처리 로직에 대한 책임을 갖고 있다면 DTO 한 곳에서만 이 로직을 책임질 수 있기에, 코드 관리와 정책(1-based)의 일관성을 가질 수 있지 않을까요? 
3. 동일한 애플리케이션이 API 마다 0-based, 1-based 등이 다르다면 사용처에서 사용하기 어렵진 않을까요?

treasure-sky replied:
1. 맞습니다. 모든 service가 이 로직을 알고 구현해야 합니다. 휴먼 에러가 발생할 가능성이 높은 것 같습니다.
2. 0-based, 1-based가 아닌 다른 방식을 사용하는 경우가 있을 수 있겠다는 생각을 해서 service에서 이를 처리하도록 했었는데, 다시 생각해보면 대부분의 경우에서 0-based, 1-based 가 사용될 거 같습니다. 따라서 DTO에서 이를 처리하는 공통 메서드를 만들어주면 일관성 측면에서나 사용성 측면에서 더 좋을 거 같습니다.
3. 네 API마다 다른 경우 지금 코드에서는 service 단에서 변환 로직을 구현해야 할텐데 위처럼 공통 메서드를 만들면 사용하기 쉬울 거 같습니다.

treasure-sky replied:
```java
public static <T> PageResponseDto<T> fromOneBasedPage(
        Page<T> page,
        int originalPageNumber
    ) {
        return new PageResponseDto<>(
            page.getContent(),
            originalPageNumber, // 원래 요청 페이지 번호
            page.getSize(),
            page.getTotalElements(),
            page.getTotalPages()
        );
    }

    public static <T> PageResponseDto<T> fromZeroBasedPage(Page<T> page) {
        return new PageResponseDto<>(
            page.getContent(),
            page.getNumber(),
            page.getSize(),
            page.getTotalElements(),
            page.getTotalPages()
        );
    }
```
그런데, DTO에 위와 같은 메서드를 추가했다고 해도 service에서 0-based, 1-based 에 대한 처리가 필요하지 않나요?
`Pageable pageable = PageRequest.of(page - 1, size, sort);` 처럼 말이죠.. 이 코드를 대체하는 방법이 떠오르지 않습니다..

hjjae2 replied:
> 그런데, DTO에 위와 같은 메서드를 추가했다고 해도 service에서 0-based, 1-based 에 대한 처리가 필요하지 않나요?

요청 흐름에서도 별도의 클래스를 만들어 사용하면 해소될 수 있지 않을까요? `PageResponseDto` 를 작성해주신 것 처럼요.

```
public record PageRequestDto<T>(
    int pageNumber,
    int pageSize
) {
    
    public PageRequest toPageRequest() {
        return PageRequest.of(pageNumber - 1, pageSize);
    }
}
```

위와 같이 한 곳에서 관리하는 모양새를 만들 수 있겠어요.

> 네 API마다 다른 경우 지금 코드에서는 service 단에서 변환 로직을 구현해야 할텐데 위처럼 공통 메서드를 만들면 사용하기 쉬울 거 같습니다.

3번 질문으로 말씀드리고 싶었던 것은, 코드 관리 측면보단 "클라이언트(프론트 애플리케이션 혹은 다른 컨트롤러, 서비스 코드 등) 측면에서 0-based, 1-based 를 둘 다 지원하는게 과연 효율적일까?" 였습니다. 

한 애플리케이션 내에 있는 서비스 코드나 API 가 0, 1 기준이 다르다면, 클라이언트는 코드, API를 사용할 때마다 어떤 기준이 사용되고 있는지 확인해야 하는 어려움이  생길텐데요, 저는 이것이 위험하고 비싼 비용이라고 생각합니다. 진호님의 생각은 어떠신가요?

treasure-sky replied:
네 requestDto를 만든다면 dto 단에서 0-based, 1-based를 관리할 수 있을 거 같습니다.
추가적으로 위에 검증 문제에 대한 답변에서 말씀드린 것처럼 dto를 사용하면 검증도 편해질 수 있을 거 같습니다.
다만, pageNumber나 pageSize를 쿼리스트링으로 보내지 않는 다는 것이 조금 걸리는데, 본문에 담아 보내도 괜찮을까요?

3번에 대한 답변으로는 두 방식을 지원하는 것이 좋지 않을 거 같다는 생각이 듭니다.
멘토님 말씀대로 하나의 애플리케이션이 여러 방식을 사용한다면 클라이언트 입장에서 혼란스러울 수 있을 거 같습니다.

hjjae2 replied:
> 다만, pageNumber나 pageSize를 쿼리스트링으로 보내지 않는 다는 것이 조금 걸리는데, 본문에 담아 보내도 괜찮을까요?

가능은 하겠습니다만, 이런 유형의 값들을 본문으로 받는 것이 일반적이라 생각하진 않아 권장드리진 않겠습니다.

인자의 타입이 DTO 클래스로 변환되면서 본문으로 받아야 된다고 말씀주신 것으로 이해했습니다.

ArgumentResolver 를 활용하여 타입/어노테이션을 탐지해 객체를 생성하는 방법도 있고([참고](https://www.baeldung.com/spring-mvc-custom-data-binder)), 현재 코드와 동일하게 외부 인자를 받고 객체를 만들어 사용하는 방법도 있겠습니다. 

```java
    @GetMapping
    public ResponseEntity<PageResponseDto<ItemResponseDto>> findAll(
        @RequestParam(defaultValue = "1")
        @Positive(message = "페이지 인덱스는 양수이어야 합니다.")
        int page,
        @RequestParam(defaultValue = "10")
        @Positive(message = "페이지 사이즈는 양수이어야 합니다.")
        int size,
        @RequestParam(defaultValue = "id") String sortBy,
        @RequestParam(defaultValue = "desc") String direction
    ) {
        var pageRequest = new PageRequestDto(page, size); // <--
        ...
```

참고로, 만약 DTO 라는 이름의 클래스를 애플리케이션 외부에서 받는 단계에서만 사용한다는 규칙이 있다면, DTO 라는 접미사를 사용하지 않고 PageRequest 와 같은 이름으로 클래스를 만들어 사용할 수도 있습니다.

hjjae2 commented on src/main/java/gift/wishlist/controller/WishlistController.java:
```diff
@@ -49,11 +59,33 @@ public ResponseEntity<WishlistResponseDto> findWishlist(
+        @Positive(message = "페이지 인덱스는 양수이어야 합니다.")
+        int page,
+        @RequestParam(defaultValue = "10")
+        @Positive(message = "페이지 사이즈는 양수이어야 합니다.")
```
항상 꼼꼼한 검증 좋네요. 👍

hjjae2 commented on src/main/java/gift/wishlist/repository/WishlistRepository.java:
```diff
@@ -13,4 +15,5 @@ public interface WishlistRepository extends JpaRepository<WishlistEntity, Long>
 
     Optional<WishlistEntity> findByIdAndMemberId(Long id, Long memberId);
 
+    Page<WishlistEntity> findByMemberId(Long memberId, Pageable pageable);
```
통상적으로 페이지네이션이란 것은 왜 필요할까요? 진호님의 생각이 궁금합니다. 🙇

treasure-sky replied:
페이지네이션을 하지 않으면 많은 문제가 발생할 수 있기 때문입니다.
1. 한 번의 조회 요청에 DB에 있는 전체 데이터를 가져와야 하는 문제가 발생합니다.
서버의 메모리 사용량이 증가하고, 쿼리 처리 시간이 증가하고, DB 병목이 발생할 수 있습니다.
2. 네트워크를 통해 사용자에게 전송하는 과정에서 과도한 비용이 발생할 수 있습니다.
3. 사용자가 저사양 기기에서 접속중이라면 대용량의 데이터를 처리하는 과정에서 서비스가 느려지거나 에러를 일으킬 수 있습니다.

hjjae2 commented on src/main/java/gift/wishlist/repository/WishlistRepository.java:
```diff
@@ -13,4 +15,5 @@ public interface WishlistRepository extends JpaRepository<WishlistEntity, Long>
 
     Optional<WishlistEntity> findByIdAndMemberId(Long id, Long memberId);
 
+    Page<WishlistEntity> findByMemberId(Long memberId, Pageable pageable);
```
Mysql DB 엔진을 사용하고 `Pageable` 을 사용하지 않는다고 가정한다면, 어떤 식의 쿼리가 필요하게 될까요?
대략적인 쿼리 예시를 제시해주실 수 있을까요-? 🙂

treasure-sky replied:
id가 22인 사용자의 위시리스트를 생성일 기준 내림차순으로 정렬하여, 첫 페이지(페이지당 10개의 위시)를 가져오는 쿼리는 다음과 같은 쿼리를 사용할 수 있을 거 같습니다.
`SELECT * FROM wishlist WHERE member_id = 22 ORDER BY created_at DESC LIMIT 10 OFFSET 0;`

OFFSET은 앞에서 얼마나 건너 뛸 건지를 나타내므로 1-based 라면 (가져올 페이지 번호 - 1) * (한 페이지당 항목 수) 로 OFFSET을 계산하여 사용할 수 있습니다.

hjjae2 commented on src/main/java/gift/wishlist/repository/WishlistRepository.java:
```diff
@@ -13,4 +15,5 @@ public interface WishlistRepository extends JpaRepository<WishlistEntity, Long>
 
     Optional<WishlistEntity> findByIdAndMemberId(Long id, Long memberId);
 
+    Page<WishlistEntity> findByMemberId(Long memberId, Pageable pageable);
```
연습하니 단계이니, 이 메서드도 테스트로 검증해볼까요? 😄

treasure-sky replied:
넵!

hjjae2 commented on src/main/java/gift/common/exception/GlobalExceptionHandler.java:
```diff
@@ -49,6 +50,37 @@ public ResponseEntity<ErrorResponseDto> handleMethodArgumentNotValidException(
+    ) {
+        ErrorResponseDto errorResponse = new ErrorResponseDto(
+            HttpStatus.BAD_REQUEST,
+            e.getMessage(),
```
이 예외의 경우 제약 사항에 대한 상세한 내용을 메시지에 포함하게 될텐데요. 시스템의 정보를 외부로 노출하는 것은 지양해야하지 않을까요?

treasure-sky replied:
네 맞습니다. 수정하도록 하겠습니다!

hjjae2 commented on src/main/java/gift/common/exception/InvalidSortDirectionException.java:
```diff
@@ -0,0 +1,8 @@
+public class InvalidSortDirectionException extends RuntimeException {
+
+    public InvalidSortDirectionException(String direction) {
+        super("올바른 정렬 방향이 아닙니다: " + direction + ", asc / desc 중 하나가 필요합니다.");
```
친절한 안내 문구네요. 좋습니다. 👍 😄

hjjae2 reviewed (changes requested):
안녕하세요, 진호님! 이번 단계, 고민과 함께 잘 구현해주셨습니다. 👍 

고민해주신 부분과 제가 궁금한 부분에 대해 몇 가지 코멘트를 남겼는데요, 확인 부탁드릴게요. 🙇 

추가로 테스트 코드가 실패하고 있는데 함께 확인해주시면 좋겠어요. 💪

treasure-sky committed:
feat: 컨트롤러에서 검증 시 예외 메세지가 너무 많은 정보를 담고 있는 문제 해결

- 기존에는 에러 메세지 전체를 반환하여 너무 많은 정보를 사용자에게 응답하였음.
- 따라서 예외에 대한 일반적인 설명만 포함하도록 하였음.

treasure-sky committed:
test: findByMemberId 테스트 추가

treasure-sky committed:
fix: 페이지네이션으로 바뀐 메서드 테스트에 반영

hjjae2 reviewed (changes requested):
안녕하세요, 진호님! 코멘트 내용 주신 잘 확인해주셨네요! 👍  

두 코멘트에 대해 추가로 질문을 남겼는데요, 이야기를 정리하고 다음 단계로 넘어갔으면 좋겠습니다. 

다시 한번 코멘트를 확인해주시고 리뷰 요청 부탁드릴게요. 🙇

treasure-sky committed:
refactor: 위시리스트 페이지네이션 요청 VO 기반 검증

- 기존 컨트롤러 단에서 이루어지던 검증을 VO와 enum 객체에서 하도록 하였음.

- 0-based, 1-based 전환을 VO와 dto에서 하도록 하였음.

treasure-sky committed:
refactor: 아이템 페이지네이션 요청 VO 기반 검증

treasure-sky committed:
refactor: 필요없는 어노테이션 삭제

treasure-sky committed:
fix: 정렬기준 key 기존 모듈에 맞게 수정

- 위시리스트에서 원래 사용하고 있던 위시리스트 id 기준 이름으로 변경 (wishlistId -> id)

treasure-sky committed:
fix: 페이지 인덱스와 크기에 대한 예외가 400을 반환하도록 수정

- 기존에는 IllegalArgumentException이 500 을 반환하였음.
- 이를 전역 예외 처리하여 400을 반환하도록 함.

treasure-sky committed:
test: 위시리스트 아이템 이름 정렬 테스트 추가

treasure-sky commented on src/main/java/gift/common/dto/PageResponseDto.java:
```diff
@@ -10,4 +11,13 @@ public record PageResponseDto<T>(
     int totalPages
 ) {
 
+    public static <T> PageResponseDto<T> from(Page<T> page) {
+        return new PageResponseDto<>(
+            page.getContent(),
+            page.getNumber() + 1, // 0-based -> 1-based
+            page.getSize(),
```
dto 단에서 0-based -> 1-based를 관리하도록 하였습니다.

treasure-sky commented on src/main/java/gift/common/vo/PageIndex.java:
```diff
@@ -0,0 +1,17 @@
+package gift.common.vo;
+
+public final class PageIndex {
+
+    private final int oneBased;
+
+    public PageIndex(int oneBased) {
+        if (oneBased < 1) {
+            throw new IllegalArgumentException("페이지 인덱스는 양수이어야 합니다. 입력값: " + oneBased);
+        }
+        this.oneBased = oneBased;
+    }
+
+    public int toZeroBased() {
+        return oneBased - 1;
+    }
+}
```
래퍼 클래스를 도입해 내부적으로 1-based 인덱스를 0-based로 관리하고, 외부에는 오직 생성자와 toZeroBased() 메서드만 노출하였습니다.
따라서 사용자가 인덱스 값을 생성하거나 조회할 때 변환 과정을 일일이 신경 쓰지 않아도, 래퍼 클래스가 자동으로 0-based로 변환해 줍니다.

treasure-sky commented on src/main/java/gift/wishlist/WishlistSortBy.java:
```diff
@@ -0,0 +1,31 @@
+import gift.common.exception.InvalidSortByException;
+import java.util.Arrays;
+
+public enum WishlistSortBy {
+    ITEM_NAME("itemName", "item.name"),
+    ITEM_PRICE("itemPrice", "item.price"),
+    CREATED_AT("createdAt", "createdAt"),
+    WISHLIST_ID("id", "id");
+
+    private final String key; // 사용자가 요청할 때 보내는 key
+    private final String property; // 실제 JPA 정렬 필드
```
아이템 정보로 위시리스트를 정렬할 때, 사용자가 요청하는 key와 실제 JPA 엔티티 필드(property)가 혼동되지 않도록, 요청용 key와 정렬용 property를 각각 분리해 관리하는 enum을 정의했습니다.

hjjae2 replied:
사소한데요, 진호님의 의도가 변수명에 잘 반영되어 보이지 않습니다. 
* `key` 가 사용자가 요청할 때 보내는 `key` 인지, jpa 에서 사용되는 `key` 인지 모호합니다.
* 하나의 예시인데요. 코틀린에선 클래스의 변수를 지칭할 때, `property` 라는 표현으로 사용하기도 합니다. 마찬가지로 의도하신 것과 다릅니다.

사용자가 요청 시 사용하는 이름 vs 인프라(DB)에서 사용하는 이름을 더 잘 표현할 수 있는 이름은 없을까요? 🤔 

예를 들면, `applicationKey` vs `infraKey` 와 같은 것 처럼요.  (꼭 이 이름을 사용하실 필요는 없습니다. 저도 만족스러운 네이밍이 막 떠오르진 않네요. 😅 )

hjjae2 reviewed (approved):
안녕하세요, 진호님! 끝까지 잘 수정해주셨네요, 고생하셨습니다. 👍 

이번 단계는 여기서 머지하도록 할게요, 다음 단계도 화이팅입니다. 💪
